[Security] Change FormAuthenticator if condition

PReimers opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30341
| License       | MIT
| Doc PR        | -

I changed the if condition in `SimpleFormAuthenticationListener` and `UsernamePasswordFormAuthenticationListener` based on the solution provided by @nikic in issue #30341 

#OpenSourceFriday

Commit:
[Security] Change FormAuthenticator if condition

PReimers commented:
Just discovered I need to write some additional tests.
They'll follow later today.

xabbuh commented on src/Symfony/Component/Security/Http/Firewall/SimpleFormAuthenticationListener.php:
```diff
@@ -107,7 +107,7 @@ protected function attemptAuthentication(Request $request)
         }
 
-        if (!\is_string($username) || (\is_object($username) && !\method_exists($username, '__toString'))) {
+        if (!\is_string($username) && !(\is_object($username) && !\method_exists($username, '__toString'))) {
```
Shouldn't it be `!\is_string($username) && (!\is_object($username) || !\method_exists($username, '__toString'))`?

PReimers replied:
You're right, the if statement will fail if it's neither a string nor an object.

I'll fix it.

PReimers commented:
I changed the if condition and added some test cases to test additional scenarios.

I hope it's no problem I added a class in the test file. This allowed me to test if the `$username` variable contains a class containing a `__toString()` method

OskarStark commented on src/Symfony/Component/Security/Http/Tests/Firewall/UsernamePasswordFormAuthenticationListenerTest.php:
```diff
@@ -99,6 +99,79 @@ public function testHandleNonStringUsername($postOnly)
+            new DefaultAuthenticationFailureHandler($this->getMockBuilder('Symfony\Component\HttpKernel\HttpKernelInterface')->getMock(), $httpUtils),
+            ['require_previous_session' => false, 'post_only' => $postOnly]
+        );
+        $event = new GetResponseEvent($this->getMockBuilder('Symfony\Component\HttpKernel\HttpKernelInterface')->getMock(), $request, HttpKernelInterface::MASTER_REQUEST);
```
you can use `createMock()` here

xabbuh replied:
No, `createMock()` does not exist in PHPUnit 4.8 which we need on our PHP 5 builds.

OskarStark replied:
Didn't see, that this targets 3.4 branch! 👍

OskarStark commented on src/Symfony/Component/Security/Http/Tests/Firewall/UsernamePasswordFormAuthenticationListenerTest.php:
```diff
@@ -99,6 +99,79 @@ public function testHandleNonStringUsername($postOnly)
+    public function testHandleNonStringUsernameWithInt($postOnly)
+    {
+        $request = Request::create('/login_check', 'POST', ['_username' => 42]);
+        $request->setSession($this->getMockBuilder('Symfony\Component\HttpFoundation\Session\SessionInterface')->getMock());
```
you should use `createMock()` here

nicolas-grekas commented:
Thank you @PReimers.
